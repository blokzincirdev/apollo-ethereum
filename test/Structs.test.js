const gql = require('graphql-tag')

const Structs = artifacts.require('Structs')

const config = {
  contracts: {
    Structs: true,
  },
}

const QUERY_CITY = gql`
  query($address: Address, $key: String) {
    Structs(address: $address) {
      city: cities(key: $key) {
        name
        state
      }
    }
  }
`

const MUTATION_ADD_CITY = gql`
  mutation($address: Address, $id: String, $name: String, $state: String) {
    Structs(address: $address) {
      createCity(id: $id, name: $name, state: $state)
    }
  }
`

contract('Structs', () => {
  let client

  before(async () => {
    client = await createClient(config, { Structs })
  })

  it('generates graphql schema', async () => {
    expect(client.source).to.matchSnapshot()
  })

  it('returns initial city', async () => {
    const variables = { address: Structs.address, key: 'sfo' }
    const data = await client.execute(QUERY_CITY, variables)
    expect(data.Structs.city).to.deep.equal({
      name: 'San Francisco',
      state: 'California',
    })
  })

  it('allows city to be added', async () => {
    await client.execute(MUTATION_ADD_CITY, {
      address: Structs.address,
      id: 'mel',
      name: 'Melbourne',
      state: 'Victoria',
    })

    const data = await client.execute(QUERY_CITY, {
      address: Structs.address,
      key: 'mel',
    })

    expect(data.Structs.city).to.deep.equal({
      name: 'Melbourne',
      state: 'Victoria',
    })
  })
})
