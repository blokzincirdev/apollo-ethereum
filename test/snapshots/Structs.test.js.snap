// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Contract: Structs generates graphql schema 1`] = `
"directive @erc1820(interfaceName: String) on FIELD_DEFINITION

scalar Address

type Mutation {
  Structs(address: Address!): StructsMutative
}

type Query {
  Structs(address: Address!): Structs
}

\\"\\"\\"
The \`Structs\` contract demonstrates the ability to use complex
objects across read and write interfaces.
\\"\\"\\"
type Structs {
  _address: Address
  cities(key: String!): Structs_City
}

type Structs_City {
  name: String
  state: String
}

\\"\\"\\"
The \`Structs\` contract demonstrates the ability to use complex
objects across read and write interfaces.
\\"\\"\\"
type StructsMutative {
  \\"\\"\\"
  Creates a new city using the provided information. If a city
  with the same ID exists, it will be overwritten
  \\"\\"\\"
  createCity(id: String, name: String, state: String): Boolean
}
"
`;
